skfem/assembly/basis/facet_basis.py

```diff
@@ -108,10 +115,110 @@ def default_parameters(self):
+    def mesh_parameters(self) -> DiscreteField:
         return DiscreteField((np.abs(self.mapping.detDG(self.X, self.find))
                               ** (1. / (self.mesh.dim() - 1.)))
                              if self.mesh.dim() != 1 else np.array([0.]))
+
+    def _trace(self,
+               x: ndarray,
+               elem: Element) -> Tuple[ndarray, ndarray, ndarray]:
+        """Trace mesh reindexing and basis projection."""
+        from skfem.utils import project
+
+        fbasis = FacetBasis(self.mesh,
+                            elem,
+                            facets=self.find,
+                            quadrature=(self.X, self.W))
+        I = fbasis.get_dofs(self.find).all()
+        if len(I) == 0:  # special case: no facet DOFs
+            if fbasis.dofs.interior_dofs.shape[0] > 1:
+                # no one-to-one restriction: requires interpolation
+                raise NotImplementedError
+            # special case: piecewise constant elem
+            I = fbasis.dofs.interior_dofs[:, self.tind].flatten()
+        y = project(x, self, fbasis, I=I)
+
+        # build connectivity for lower dimensional mesh
+        ix = self.mesh.facets[:, self.find]
+        ixuniq = np.unique(ix)
+        b = np.zeros(np.max(ix) + 1, dtype=np.int64)
+        b[ixuniq] = np.arange(len(ixuniq), dtype=np.int64)
+
+        return self.mesh.p[:, ixuniq], b[ix], y
+
+    def trace(self,
+              x: ndarray,
+              projection: Callable[[ndarray], ndarray],
+              target_elem: Optional[Element] = None) -> Tuple[InteriorBasis,
+                                                              ndarray]:
+        """Restrict solution to :math:`d-1` dimensional trace mesh.
+
+        The user must define how the boundary points are projected using the
+        argument ``projection``.  For example,
+
+        >>> projection = lambda p: p[0]
+
+        will keep only the `x`-coordinate when initializing the trace mesh.
+
+        Parameters
+        ----------
+        x
+            The solution vector.
+        projection
+            A function defining the projection of the boundary points.  See
+            above for an example.
+        target_elem
+            Optional lower-dimensional finite element to project to.  If not
+            given, a piecewise constant element is used.
+
+        Returns
+        -------
+        InteriorBasis
+            An object corresponding to the trace mesh.
+        ndarray
+            A projected solution vector defined on the trace mesh.
+
+        """
+        DEFAULT_TARGET = {
+            MeshTri: ElementLineP0,
+            MeshQuad: ElementLineP0,
+            MeshTet: ElementTriP0,
+            MeshHex: ElementQuad0,
+        }
+
+        meshcls = type(self.mesh)
+        if meshcls not in DEFAULT_TARGET:
+            raise NotImplementedError("Mesh type not supported.")
+        if target_elem is None:
+            target_elem = DEFAULT_TARGET[meshcls]()
```

Comment:
Would it be more Pythonic to have the `DEFAULT_TARGET` as an attribute or property of `self.mesh`?  And EAFP over LBYL?  Define as `NotImplemented` or similar for `Mesh` then override in subclasses?  The resulting default error message should be almost as obvious.

Could the key use `self.mesh.brefdom`? 
```python
{'line': ElementLineP0, 'tri': ElementTriP0, 'quad': ElementQuad0}
```
(and worry about wedges #411 &c. later&hellip;)